### ISSUE TITLE
Filters in Queries

### ISSUE THREAD

**Issue by PetrSvirak** [MacakM/PV247-Expense-Manager#42]
_GetQueryable_ override is almost always way too complex method. Also, it does not make sense for the query class itself to implement individual filters. Any change in the filter class leads to change in the query, meaning elementary SOLID principles are broken here. Come up with a better solution and facilitate generally same behavior more properly.

**Comment by PetrSvirak** [comment]
This does not solve the problem, does it? Now, the _FilterQuery_ method is too complex. It needs to be more granular, the method is (still) too much of a monolith.

**Comment by PetrSvirak** [comment]
Well, first some more or less details: class ExpenseManager.Business.DataTransferObjects.Filters.Filter is not an interface while comment state so. Also, even it was an interface, it is a sign of wrong design when an interface is empty. This interface is not only empty, it is also generic, making it even worse.

While this is definitely a good approach to define filters using  ExpenseManager.Database.Filters.FilterModel<T> (btw: should be an interface as well), there seem to be a bit more filters than actually needed. Overall, the implementation seems too much universal in terms of predefined filters. The design here is though correct, I like the idea itself!

I really dislike the way filters are pushed around through whole application, though. In the current design, the database objects are accessible in the business layer, thus there is no need for the FilterModel and its implementations to be in the database layer. Presentation layer should not know anything about these filters at all. It should only pass arguments as scalars to the service (facades in your case) or as a simple POCO object published by/based on given service to separate individual groups of service's method arguments. In your case, I would expect facades to be responsible for the conversion of scalars to filters.

In other words: All filters implementations should be internal. No public filter is required as it is a services' job to amend queries accordingly to the presentation layer request and there is no need for them to exist in the database layer either.

Here we leverage publicity of DB models in services API. It is somewhat in nature of EF framework, or more precisely, queryable interfaces that they need to make data contract public as conversion causes materialization or query incompatibility. We can go with single service layer objects implementation (where service layer provides a contract to the DB layer and the later cares about conversions), however, this would require a considerable amount of business logic (including filters) moved to the DB layer. It is up to you though.

**Comment by PetrSvirak** [comment]
One more thing: It is not as important as you will be removing the auto-mapper configuration anyway, however, with this number of implementations, it would be better to use reflexion and dynamically register individual implementations of the interfaces. This way, one can avoid mapping some implementations twice  (or eventually forgetting some). If you did not like the idea of reflexion in production code (while you (as a team) use in on several places), there should at least be a test that ensures all implementations are correctly registered.

**Comment by PetrSvirak** [comment]
First of all: Definitely no tuples.
Second: there is no reason for all this reflexion machinery. Simply create new instances of filters in the FilterFactory. Also, do not use default constructor and public properties. In this case, filters should clearly have constructors with the parameter value they filter on. Property can be made read-only (having only getter and initialized in the constructor).
Third: Definitely no tuples.

**Comment by PetrSvirak** [comment]
Almost. Now you returned the responsibility of checking the value back to the individual filter. That is completely unnecessary.

First, when I wrote about the read-only property, I meant it. The syntax is as follows:
````
public Guid AccountId { get; }
````
It is perfectly reasonable for a filter to expose the value it is filtering on. And it would be essential for tests and debug.

Second, there is no addition to any filter lists you use. It is actually not meant for addition by design, isn't  it? There is thus no reason to use ``List<T>``. Even if there was any, I would still encourage you to use ``IList<T>`` unless there is a good reason for the concrete type to be used. But in this case, you should not even assume that all filter will have to be evaluated (e.g. your query could be a bit smarter and stop processing filters when there are no items matching to the filters that were already applied. Replace all ``List<T>`` return types in the factory with simple ``IEnumerable<T>``. This allows you to replace your return ``new List<T>`` { …  } that spread over the factory with simple rows all starting with ``yield return``.

Third, there are at least two nice options you seem to overlooked. 
1. Generic value interface
There could an interface like this
````
internal interface IFilterValue<in V>
 {
     V Value { set; }
 }
````
Each filter will then implement this interface. For example ``AccountsByName`` could look like this:
````
internal class AccountsByName : IFilter<AccountModel>, IFilterValue<string>
{
    public string Value { set; private get;  }

    public IQueryable<AccountModel> FilterQuery(IQueryable<AccountModel> queryable)
        => return queryable.Where(account => account.Name.Contains(Value));
}
````
Then in the factory there could be 2 methods like this:
````
private static TFilter TryCreateFilter<TFilter, TValue>(TValue value)
    where TFilter : class, IFilterValue<TValue>, new()
{
    return value == null ? new TFilter() { Value = value } : null;
}

private static TFilter TryCreateFilter<TFilter, TValue>(TValue? value)
    where TFilter : class, IFilterValue<TValue>, new()
    where TValue : struct
{
    return value.HasValue ? new TFilter() { Value = value.Value } : null;
}
````
The first method allows all types as far as the filter expects the same value as provided. If ``TValue`` is a value type, it will never be ``null``, so filter is always created. If ``TValue`` is reference type, filter is only created for not-``null`` values.
The second method is meant for value types only, accepting their nullable alternatives and thus allowing even these filters not to be created if there was no value provided to the value-type wrapper.

Usage would look like this:
````
yield return TryCreateFilter<AccountBadgesByAccountId, Guid>(accountId);
````

2. Generic creation function
Filters can stay almost similar in the way they are (remove the condition and see the read-only properties above). For example, ``CostInfosByAccountId`` could look like this:
````
internal class CostInfosByAccountId : IFilter<CostInfoModel>
{
    public Guid AccountId { get; }

    public CostInfosByAccountId(Guid accountId)
    {
        AccountId = accountId;
    }

    public IQueryable<CostInfoModel> FilterQuery(IQueryable<CostInfoModel> queryable)
        => queryable.Where(x => x.AccountId == AccountId);
}
````

However, the factory would be equipped with 2 methods similar to these:
````
private static TFilter TryCreateFilter<TFilter, TValue>(TValue value, Func<TValue, TFilter> filterConstructor)
    where TFilter : class, IFilter<TValue>
{
    return value == null ? null : filterConstructor(value);
}

private static TFilter TryCreateFilter<TFilter, TValue>(TValue? value, Func<TValue, TFilter> filterConstructor)
    where TFilter : class, IFilter<TValue>
    where TValue : struct
{
    return value.HasValue ? filterConstructor(value.Value) : null;
}
````
Methods meaning is very similar to the ones in the first option. The second method only applies to value types, while the first method works with both values and references.

This would allow you to yield filters in individual public factory methods like this:
````
yield return TryCreateFilter(accountId, id => new CostInfosByAccountId(id));
````

While the first option might seem a bit more complex, it is also more generic in the very meaning of the word. It allows a base class to be extracted for all filters. It could look something like that:
````
internal abstract class FilterValueBase<TEntity, TValue> : IFilter<TEntity>, IFilterValue<TValue>
{
    public TValue Value { set; private get; }

    public abstract Expression<Func<TEntity, bool>> GetWhereCondition(TValue value);

    public IQueryable<TEntity> FilterQuery(IQueryable<TEntity> queryable)
        => queryable.Where(GetWhereCondition(Value));
}
````
Then specifying individual filters would really be very simple. For example ``CostInfosByCreatedFrom`` could look like this:
````
internal class CostInfosByCreatedFrom : FilterValueBase<CostInfoModel, DateTime>
{
    public override Expression<Func<CostInfoModel, bool>> GetWhereCondition(DateTime value)
        => costInfo => costInfo.Created >= value;
}
````

**Note** that in both cases there is only one place where the ``null`` condition is evaluated and where structures are handled (single point of failure, responsibilities split). Also, you should notice that filters' enumerable can contain null filters - meaning filtering by given value does not make sense in given request. This needs to be handled either 
1. by the end-enumerable consumer, in your case, it is the ``ExpenseManagerQuery.ApplyFilters`` method:
````
queryable = Filters.Aggregate(queryable, (current, filter) => filter == null ? current : filter.FilterQuery(current));
// or similar but more expressive
queryable = Filters.Where(filter => filter != null).Aggregate(queryable, (current, filter) => filter.FilterQuery(current));
````
2. by using [Null-object pattern](https://sourcemaking.com/design_patterns/null_object).

The first option si a bit more suitable for this pattern, so I will use it.
Creating the null object could be really easy. Class defining it could like this:
````
private class NullFilter<TEntity, TValue> : FilterValueBase<TEntity, TValue>
{
    public override Expression<Func<TEntity, bool>> GetWhereCondition(TValue value) => _ => true;
}
````
The helper methods in the factory would have to be modyfied a bit:
````
private static IFilter<TEntity> TryCreateFilter<TEntity, TValue>(TValue value, Func<TValue, IFilter<TEntity>> filterConstructor)
{
    return value == null
        ? new NullFilter<TEntity, TValue>()
        : filterConstructor(value);
}

private static IFilter<TEntity> TryCreateFilter<TEntity, TValue>(TValue? value, Func<TValue, IFilter<TEntity>> filterConstructor)
    where TValue : struct
{
    return value.HasValue
        ? filterConstructor(value.Value)
        : new NullFilter<TEntity, TValue>();
}
````
This way, there would be no ``null``s in the filters' enumerable of any public factory method.
Of course, it is not ideal to always create a new instance of supposedly singleton object, yet what you see here is but a sample code to give you the general idea.

**Comment by xgasior** [comment]
ok?

**Comment by PetrSvirak** [comment]
There are several things I don't like:
* methods are mostly called with plenty of ``null`` arguments. There are 3 solutions to this: multiple overrides, named arguments with a default value, POCO object carrying arguments within, with multiple constructors.
* it seems that the code was written in quite a haste. Your own coding rules are broken here and there.
* copy-pasted parts of the code are visibly different to the rest of your project; especially because at some places the code does not benefit from C# 6 features while it does in the pasted code.

While these are quite essential for any production API maintained by multiple developers, this is still a school project. You have done quite work on this task. Hopefully, you can still see advantages of this approach. Let's close it with that.

**Comment by xgasior** [comment]
Ok, thanks.

I will fix 1st and 2nd point before redeploy and i will close it.
